### ISSUE TITLE
Problem with OCR similarity map

### ISSUE THREAD

**Issue by Mayy1994** [openseg-group/openseg.pytorch#22]
Thanks for sharing this wonderful work with us!

I have a problem with the computing of similarity map in the OCR module.
In **line 131** in lib/models/seg_hrnet_orc.py
        `sim_map = (self.key_channels**-.5) * sim_map` 
Why multiply a small value (`self.key_channels**-.5`) to sim_map before softmax?

During validation, I have printed the final result of `sim_map` and I found all values in this map are very close to 0.0526 (equals to 1/19), which means the probabilities of a pixel `i` belong to different classes` k` are almost equal.
Is this contradicting the assumption that the similarity map should represent the relation between the _i_th pixel and the _k_th object region?

#######################

Your former answer:

- Multiplying the small value is following the original self-attention scheme. Please refer to the last paragraph of 3.2.1 in the paper "Attention Is All You Need". However, we find this small factor does not influence the segmentation performance.

- As the final result of the sim_map, we do not understand why all the values are almost the same in your case. What checkpoints are you testing? How about the performance of the used checkpoint? Please provide more information so that we can help you.

#########################

Thanks a lot for your reply!
I used the checkpoint posted on HRNet-OCR. The segmentation performance is good ad the mIoU is 81.6, too.
![Screenshot from 2020-06-15 12-07-11](https://user-images.githubusercontent.com/40232175/84617296-10ce3900-af01-11ea-8477-7aa922109cc6.png)
In inference, I have printed 10 random rows in the `sim_map` like below:
![Screenshot from 2020-06-15 12-19-24](https://user-images.githubusercontent.com/40232175/84617764-8e467900-af02-11ea-923f-b76e933a08c8.png)
All values in this map are very close to 0.0526 (equals to 1/19).

**Comment by PkuRainBow** [comment]
We recommend you to sample pixels in a more reasonable manner instead of fully random.  We can not infer reliable conclusions only based on fully random sampled pixels.

For example, you can randomly sample the pixels that belong to the person or the pixels that belong to the car. 

In fact, this observation (based on the randomly sampled positions) is not surprising and we will also check this problem.

**Comment by Mayy1994** [comment]
Yes, I also randomly sampled some pixels that belong to the 19 classes of Cityscapes, e.g. road, building, vegetation, car, but the values are still close to 0.0526.
If you have any further findings for this, pls let me know. Many thanks.

**Comment by PkuRainBow** [comment]
@Mayy1994 We recommend you to print the sim_map = (self.key_channels**-.5) * sim_map before softmax normalization or print the sim_map w/o dividing the self.key_channels.

**Comment by Mayy1994** [comment]
I printed some pixels in the sim_map w/o dividing the self.key_channels and w/o softmax normalization. 
This is the picture of the pixels I choose:
<img width="946" alt="Screenshot 2020-06-15 at 5 19 18 PM" src="https://user-images.githubusercontent.com/40232175/84640566-7126a000-af2c-11ea-92c5-8ef6c37285b3.png">
This is the sim_map values of the corresponding pixels. However, different pixels have the same biggest similarity with the Class Car (index: 13).
<img width="629" alt="Screenshot 2020-06-15 at 5 09 42 PM" src="https://user-images.githubusercontent.com/40232175/84639644-3bcd8280-af2b-11ea-8233-fdc34bff99bc.png">
I also printed the max value of all pixels in the class dimension. Here is the result:
![Screenshot from 2020-06-15 16-51-24](https://user-images.githubusercontent.com/40232175/84640001-b1d1e980-af2b-11ea-8e31-bed6f54c9d42.png)
So weird! It is really difficult to understand.

**Comment by PkuRainBow** [comment]
Both the road pixel and the wall pixel choose the **car object representation** seems very unreasonable. 

It would be great if you could share with more details of the testing code and we would like to help to check the possible reasons.

@hsfzxjy

**Comment by Mayy1994** [comment]
I just used the official /tools/test.py for validation. By loading the same checkpoint posted in HRNet-OCR, I think you can obtain the same results.

I tested several images in the validation dataset, but a similar situation occurred. The pixels may choose a wrong object representation.

**Comment by PkuRainBow** [comment]
Thanks for your feedback. We will investigate the possible reasons and share our conclusions with you in the next few weeks.

**Comment by Mayy1994** [comment]
Many thanks. Looking forward to hearing from you at your earliest convenience.

**Comment by PkuRainBow** [comment]
@Mayy1994 We have checked the checkpoints of this code base and we do not find the mentioned problems. 

Currently, we encourage you to verify the similarity maps based on the openseg.pytorch code base. @hsfzxjy will help to share more detailed results later.

**Comment by hsfzxjy** [comment]
hi @Mayy1994 . We have confirmed the phenomenon you mentioned in repo HRNet-Semantic-Segmentation. The maximum of similarity value of an image are gathering in very few categories.  There may be some impelementation mistakes in it.

Besides, we also do the same test on this repo (openseg.pytorch), and the problem does not occur.
Specifically, we follow your code `torch.max(sim, dim=-1)[1].unique()` to print out the maximum locations. The results are presented in the screenshot below.
![image](https://user-images.githubusercontent.com/4702188/84655662-eea9da80-af43-11ea-8934-f484374eb56f.png)

We encourage you to verify the issue based on this repo. We will figure out the problem in HRNet-Semantic-Segmentation as soon as possible.

**Comment by Mayy1994** [comment]
@hsfzxjy Thanks for your sharing. I will test it based on this repo.
